File: benchmarks/compare/dijkstra/dijkstra.go
// dijkstra.go — linear-scan priority-queue Dijkstra (O(V^2)) over an in-code graph.
// Graph generated in memory by a portable deterministic generator (no file I/O).
// Prints exactly one stdout line "RESULT=<int>". Generator: Park-Miller MINSTD.
package main

import "fmt"

const n = 30000
const INF = 1000000000

type edge struct{ to, w int }

func main() {
	state := int64(1234)
	nxt := func() int64 { state = (state * 16807) % 2147483647; return state }

	adj := make([][]edge, n)
	for i := 0; i < n; i++ {
		for d := 1; d <= 8; d++ {
			j := i + d
			if j < n {
				w := int(nxt()%100) + 1
				adj[i] = append(adj[i], edge{j, w})
			}
		}
		if i+1 < n {
			r := nxt()
			if r%10 < 3 {
				span := int64(n - (i + 1))
				j := (i + 1) + int(nxt()%span)
				w := int(nxt()%100) + 1
				adj[i] = append(adj[i], edge{j, w})
			}
		}
	}

	dist := make([]int, n)
	visited := make([]bool, n)
	for i := range dist {
		dist[i] = INF
	}
	dist[0] = 0
	capacity := n*9 + 1
	pqn := make([]int, capacity)
	pqd := make([]int, capacity)
	pql := 1
	for pql > 0 {
		mi := 0
		for j := 1; j < pql; j++ {
			if pqd[j] < pqd[mi] {
				mi = j
			}
		}
		u := pqn[mi]
		last := pql - 1
		pqn[mi] = pqn[last]
		pqd[mi] = pqd[last]
		pql = last
		if !visited[u] {
			visited[u] = true
			du := dist[u]
			for _, e := range adj[u] {
				nd := du + e.w
				if nd < dist[e.to] {
					dist[e.to] = nd
					pqn[pql] = e.to
					pqd[pql] = nd
					pql++
				}
			}
		}
	}
	var total int64
	for k := 0; k < n; k++ {
		if dist[k] < INF {
			total += int64(dist[k])
		}
	}
	chk := int64(dist[n-1])*1000003 + (total % 1000000000)
	fmt.Printf("RESULT=%d\n", chk)
}

File: benchmarks/compare/dijkstra/dijkstra.rs
// dijkstra.rs — linear-scan priority-queue Dijkstra (O(V^2)) over an in-code graph.
// Graph generated in memory by a portable deterministic generator (no file I/O).
// Prints exactly one stdout line "RESULT=<int>". Generator: Park-Miller MINSTD.
const N: usize = 30000;
const INF: i64 = 1000000000;

fn main() {
    let mut state: i64 = 1234;
    let mut nxt = || {
        state = (state * 16807) % 2147483647;
        state
    };

    let mut adj: Vec<Vec<(usize, i64)>> = vec![Vec::new(); N];
    for i in 0..N {
        for d in 1..=8 {
            let j = i + d;
            if j < N {
                let w = nxt() % 100 + 1;
                adj[i].push((j, w));
            }
        }
        if i + 1 < N {
            let r = nxt();
            if r % 10 < 3 {
                let span = (N - (i + 1)) as i64;
                let j = (i + 1) + (nxt() % span) as usize;
                let w = nxt() % 100 + 1;
                adj[i].push((j, w));
            }
        }
    }

    let mut dist = vec![INF; N];
    let mut visited = vec![false; N];
    dist[0] = 0;
    let capn = N * 9 + 1;
    let mut pqn = vec![0usize; capn];
    let mut pqd = vec![0i64; capn];
    let mut pql = 1usize;
    while pql > 0 {
        let mut mi = 0;
        for j in 1..pql {
            if pqd[j] < pqd[mi] {
                mi = j;
            }
        }
        let u = pqn[mi];
        let last = pql - 1;
        pqn[mi] = pqn[last];
        pqd[mi] = pqd[last];
        pql = last;
        if !visited[u] {
            visited[u] = true;
            let du = dist[u];
            for k in 0..adj[u].len() {
                let (v, w) = adj[u][k];
                let nd = du + w;
                if nd < dist[v] {
                    dist[v] = nd;
                    pqn[pql] = v;
                    pqd[pql] = nd;
                    pql += 1;
                }
            }
        }
    }
    let mut total: i64 = 0;
    for k in 0..N {
        if dist[k] < INF {
            total += dist[k];
        }
    }
    let chk = dist[N - 1] * 1000003 + (total % 1000000000);
    println!("RESULT={}", chk);
}
